packages/framework/src/Model/Product/Recalculation/ProductRecalculationFacade.php

```diff
@@ -34,14 +38,75 @@ public function __construct(
      */
     public function recalculate(array $productIds): void
     {
-        $this->productVisibilityFacade->calculateProductVisibilityForIds($productIds);
+        $idsToRecalculate = $this->getIdsToRecalculate($productIds);
+
+        $this->productVisibilityFacade->calculateProductVisibilityForIds($idsToRecalculate);
 
         foreach ($this->domain->getAllIds() as $domainId) {
             $this->indexFacade->exportIds(
                 $this->indexRegistry->getIndexByIndexName(ProductIndex::getName()),
                 $this->indexDefinitionLoader->getIndexDefinition(ProductIndex::getName(), $domainId),
-                $productIds,
+                $idsToRecalculate,
             );
         }
     }
+
+    /**
+     * @param array $productIds
+     * @return array
+     */
+    protected function getIdsToRecalculate(array $productIds): array
+    {
+        $mainVariantsIds = $this->getMainVariantIdsOfVariants($productIds);
+
+        $productIdsWithAddedMainVariantIds = [...$productIds, ...$mainVariantsIds];
+
+        $regularProductOrMainVariantIds = $this->dropVariantIdsFromProductIds($productIdsWithAddedMainVariantIds);
+        $variantIds = $this->getVariantIds($productIdsWithAddedMainVariantIds);
+
+        return [...$variantIds, ...$regularProductOrMainVariantIds];
+    }
+
+    /**
+     * @param int[] $productIds
+     * @return int[]
+     */
+    protected function getMainVariantIdsOfVariants(array $productIds): array
+    {
+        $result = $this->em->createQuery('SELECT IDENTITY(p.mainVariant) as id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND p.variantType = :variantTypeVariant')
+            ->setParameter('productIds', $productIds)
+            ->setParameter('variantTypeVariant', Product::VARIANT_TYPE_VARIANT)
+            ->getResult();
+
+        return array_column($result, 'id');
+    }
+
+    /**
+     * @param int[] $productIds
+     * @return int[]
+     */
+    protected function dropVariantIdsFromProductIds(array $productIds): array
+    {
+        $result = $this->em->createQuery('SELECT p.id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND (p.variantType = :variantTypeMain OR p.variantType = :variantTypeNone)')
+            ->setParameter('productIds', $productIds)
+            ->setParameter('variantTypeMain', Product::VARIANT_TYPE_MAIN)
```

Comment:
I am wondering which approach is safer:
```
p.variantType = :variantTypeMain OR p.variantType = :variantTypeNone
```
versus
```
p.variantType != :variantTypeVariant
```

Maybe rather the second one? I mean, if someone adds another variant type, they will need to add the type here with the first option, right?

This is really hypothetical, I believe we will be ok even with the first solution, I do not remember anyone adding another variant type, so :shrug:

Reply:
your proposal is easier to read and make sense to me (based on the method name), so changed